compiler/rustc_codegen_ssa/src/back/linker.rs

```diff
@@ -1691,6 +1697,26 @@ pub(crate) fn exported_symbols(tcx: TyCtxt<'_>, crate_type: CrateType) -> Vec<St
     symbols
 }
 
+fn exported_symbols_for_proc_macro_crate(tcx: TyCtxt<'_>) -> Vec<String> {
+    let mut symbols = Vec::new();
+
+    let stable_crate_id = tcx.sess.local_stable_crate_id();
+    let proc_macro_decls_name = tcx.sess.generate_proc_macro_decls_symbol(stable_crate_id);
+    let metadata_symbol_name = exported_symbols::metadata_symbol_name(tcx);
+
+    // You would think that both the two names would always be there, but in
+    // pnkfelix's local experiments that was not case. So instead we walk the
+    // list and only add them if they *are* there.
```

Comment:
Surprising. In what kind of cases is it missing?

Reply:
I didn't investigate too deeply; my memory was that it didn't pass the test suite, so I switched to this...

... but actually, now that I say that: My test suite was borked (without my realizing it) when I was first investigating this (due to my wacky [columns env var issue](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/rustc_expand.20test.20failures.20on.20nightly/near/377370244)). I should re-do the experiment again now that that is resolved.

(I do think that the failing tests were clearly related to this problem; the columns thing was only in later test phases. But it is still worth teasing out explicitly.)

Reply:
Okay, so I made a variation that emits warnings when each of the two symbols for a proc-macro crate are missing. 

Here are the first set of ui test suite failures with that in place, all three of which emit neither the decls list symbol nor the metadata symbol.

```
failures:
    [ui] tests/ui/proc-macro/allowed-signatures.rs
    [ui] tests/ui/proc-macro/panic-abort.rs
    [ui] tests/ui/proc-macro/proc-macro-deprecated-attr.rs
```

All three of these are marked as "check-pass" (and I'm now wondering if my own test in this PR should have been "check-pass" instead of "build-pass", which very few of the other proc-macro tests have ...)

I skimmed their sources (linked below) but nothing immediately struck me as being a common root cause that would make *these* specific files fall into the cases in question.

[allow-signatures.rs](https://github.com/rust-lang/rust/blob/5881e5f88d9245ef9259ca600b32af80d5972a7f/tests/ui/proc-macro/allowed-signatures.rs)

[panic-abort](https://github.com/rust-lang/rust/blob/5881e5f88d9245ef9259ca600b32af80d5972a7f/tests/ui/proc-macro/panic-abort.rs)

[proc-macro-deprecated-attr.rs](https://github.com/rust-lang/rust/blob/5881e5f88d9245ef9259ca600b32af80d5972a7f/tests/ui/proc-macro/proc-macro-deprecated-attr.rs)

Reply:
> and I'm now wondering if my own test in this PR should have been "check-pass" instead of "build-pass", which very few of the other proc-macro tests have ...

I don't think `check-pass` on this test would truly represent a regression test as `check` doesn't codegen or link the crate.

Reply:
Oh wait, ... I think I found the common elements here.

I hypothesize the three test crates listed in my previous comment are the *only* crates in the ui/proc-macro test directory itself (i.e. not the auxiliary directory) that are:
 1. *proc-macro* crates (as opposed to being *users of proc-macros*),
 2. that do not themselves import `extern crate proc_macro;`, and
 3. are set as "check-pass" (rather than "compile-fail") 
 
*That* might explain somewhat why they're singled out in their behavior here.

And maybe these are cases we should not be concerned about supporting? I.e., any well-behaved proc-macro crate should be including `extern crate proc_macro;` in its source listing, right?

Reply:
(Also there's been some relatively recent movement here that might be relevant, in terms of making assumptions about whether codegen does or does not need to be exercised, in PR #114124 )

Reply:
For a check-pass build the linker code shouldn't be invoked at all, right?

Reply:
> For a check-pass build the linker code shouldn't be invoked at all, right?

The linker wouldn't be invoked, right. But my current understanding is that the `exported_symbols` procedure is still *run*, and thus the code I added to this PR to perform the [aforementioned experiment ](https://github.com/rust-lang/rust/pull/114470#discussion_r1284877313) is still exercised (and so we end up seeing the diagnostics I'm emitting, and so the tests fail.

Reply:
Here is how I see things. The choice I have in this PR is either:

 1.  unconditionally emit the two symbols, or
 2.  conditionally emit them if and only if we find in them in the traversal performed by `for_each_exported_symbols_include_dep`.

To me, it seems wrong to emit the two symbols if they aren't part of the `for_each_exported_symbols_include_dep` traversal.

So in my head, we can recast the two options above as:

 1. There *should* be an invariant that the two symbols are always in the `for_each_exported_symbols_include_dep` traversal (which would justify the unconditional emission), but today, if that invariant is supposed to hold, it clearly does not. Or
 2. There is no such invariant, and so the emission must be conditional (in the manner that this PR is written today).

Reply:
I suppose I could post a pernos.co build of one of the cases showing the invariant failing. I'll plan to do that on Monday.

Reply:
Does the invariant ever fail in non-check mode? I think `for_each_exported_symbols_include_dep` simply returns no symbols at all in check mode. It would make sense to me to check `tcx.sess.opts.output_types.should_codegen()` and return an empty list if false.

Reply:
Okay I'll investigate that option.

Reply:
> It would make sense to me to check tcx.sess.opts.output_types.should_codegen() and return an empty list if false.

This is now done, in commit https://github.com/rust-lang/rust/pull/114470/commits/a2058ddbed82adda55ce39f0e0915996706cb2b9